draft-ietf-mls-protocol.md

```diff
@@ -494,26 +502,105 @@ as follows:
 used for the Merkle trees in the Certificate Transparency protocol
 {{?I-D.ietf-trans-rfc6962-bis}}.)
 
-## Ratchet Tree Nodes and Tree Updates
+The leaves of the tree are indexed separately, using a _leaf index_,
+since the protocol messages only need to refer to leaves in the
+tree.  Like nodes, leaves are numbered left to right.  Note that
+given the above numbering, a node is a leaf node if and only if it
+has an even node index, and a leaf node's leaf index is half its
+node index.  The leaf indices in the above tree are as follows:
+
+* 0 = A
+* 1 = B
+* 2 = C
+* 3 = D
+* 4 = E
+* 5 = F
+* 6 = G
+
+
+## Ratchet Tree Nodes
 
-Ratchet trees are used for generating shared group secrets. In this
-section, we describe the structure of a ratchet tree.  A particular
-instance of a ratchet tree is based on the following cryptographic
-primitives, defined by the ciphersuite in use:
+A particular instance of a ratchet tree is based on the following
+cryptographic primitives, defined by the ciphersuite in use:
 
 * A Diffie-Hellman finite-field group or elliptic curve
 * A Key Derivation Function (KDF)
 * A Derive-Key-Pair function that produces an asymmetric keypair
   from a node secret
 
-A ratchet tree is a left-balanced binary tree, in which each node
-contains up to three values:
+Each node in a ratchet tree contains up to three values:
+
+* A private key (only within direct path, see below)
+* A public key
+* A credential (only for leaf nodes)
+
+The conditions under which each of these values must or must not be
+present are laid out in {{views}}.
+
+A node in the tree may also be _blank_, indicating that no value is
+present at that node.  The _resolution_ of a node is an ordered list
+of non-blank nodes that collectively cover all non-blank descendants
+of the node.  The nodes in a resolution are ordered according to
+their indices.
+
+* The resolution of a non-blank node is a one element list
+  containing the node itself
+* The resolution of a blank leaf node is the empty list
+* The resolution of a blank intermediate node is the result of
+  concatinating the resolution of its left child with the resolution
+  of its right child, in that order
```

Comment:
Editorial: "in that order" seems redundant or meaningless here

Reply:
To me "in that order" somewhat clarifies that "left.child || right.child" is meant here and not "right.child || left.child". Clear yet though would be something like: 

The resolution of a blank intermediate node is the result of concatenating the resolution of its right child to the end of the resolution of its left child.